### DISCUSSION TITLE
How to observe the inovke of command in other viewmodel?

### DISCUSSION THREAD

**Discussion by Snailya** [reactiveui/ReactiveUI#3779] [Q&A]
I have a ViewModel name MainViewModel which host another ViewModel name SubViewModel as Property. In the SubViewModel, there is a command named Select. I want to observe this command invoke in the MainViewModel to do some side effect.
I tried this in MainViewModel's constructor, but not workd.
`
SubViewModel.WhenAnyObservable(x=>x.Select).Subscribe(x=>{//todo});
`

**Comment by dpvreony** [comment]
it depends what you're trying to observe

if you want to just observe if the command can or is executing you can use the following observables (also showing how to watch for exceptions)
```
            SubViewModel.Select.CanExecute.Subscribe(x => OnCanExecute(x));
            SubViewModel.Select.IsExecuting.Subscribe(x => OnIsExecuting(x));
            SubViewModel.Select.ThrownExceptions.Subscribe(x => OnThrownException(x));
```

if you're wanting the output of the command, it "can" be "easier" (if it's a read only value that gets pulled from the command and isn't changed once in the view model) to push into an observable property first. in the constructor of your subviewmodel

```
   public sealed class SubViewModel : ReactiveUI.ReactiveObject
   {
       private readonly ObservableAsPropertyHelper<string> _somePropertyObservable;

       public SubViewModel()
       {
            // you would want a better can execute logic, this is minimal example
            var canExecute = Observable.Return(true);
            Select = ReactiveCommand.CreateFromTask(_ => "here", canExecute);
            _somePropertyObservable = Select.ToProperty(
                this,
                vm => vm.SomeProperty,
                scheduler: RxApp.MainThreadScheduler);
       }

       public string SomeProperty => _somePropertyObservable.Value;

       public ReactiveCommand<Unit, string> Select
       {
           get;
       }
   }           
```

then you want the following in your main view model
```
   this.WhenAnyValue(this, vm => vm.SubViewModel.SomeProperty).Subscribe(x => ...);
```

you could use `this.SubViewModel.Select.Subcribe(x => ...)` instead of the observable property helper, it depends on lifespans of subviewmodels, if you're already binding to the data out of the command into the subviewmodel and doing 2 way binding to the UI, and how you want to structure code.
